core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -336,15 +323,15 @@ public T acquire(long time, TimeUnit unit)
    */
   @Override
   public void release(T resource) {
+    if (!mResources.containsKey(resource)) {
+      throw new IllegalArgumentException(
+          "Resource " + resource.toString() + " was not acquired from this resource pool.");
+    }
+    ResourceInternal<T> resourceInternal = mResources.get(resource);
```

Comment:
between the previous `containsKey` and this `get` it could be removed and this could be null. This has to be coordinated with the removal (probably by the `mLock`).